NullPointerException when refreshing location

1. Settings -> User Interface -> Show Moon .. uncheck this option
2. Set Location -> Current (last known) -> update

The app crashes with the following exception:
```
2020-05-03 00:11:17.715 473-473/com.forrestguice.suntimeswidget E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.forrestguice.suntimeswidget, PID: 473
    java.lang.NullPointerException: Attempt to invoke virtual method 'void com.forrestguice.suntimeswidget.calculator.SuntimesMoonData.invalidateCalculation()' on a null object reference
        at com.forrestguice.suntimeswidget.SuntimesActivity.invalidateData(SuntimesActivity.java:1449)
        at com.forrestguice.suntimeswidget.SuntimesActivity$8.onStart(SuntimesActivity.java:840)
        at com.forrestguice.suntimeswidget.getfix.GetFixTask.onPreExecute(GetFixTask.java:165)
        at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:648)
        at com.forrestguice.suntimeswidget.getfix.GetFixTask.executeTask(GetFixTask.java:64)
        at com.forrestguice.suntimeswidget.getfix.GetFixHelper.getFix(GetFixHelper.java:126)
        at com.forrestguice.suntimeswidget.SuntimesActivity$17.onClick(SuntimesActivity.java:1244)
        at com.forrestguice.suntimeswidget.LocationConfigDialog$4.onClick(LocationConfigDialog.java:324)
        at android.view.View.performClick(View.java:6597)
        at android.view.View.performClickInternal(View.java:6574)
        at android.view.View.access$3100(View.java:778)
        at android.view.View$PerformClick.run(View.java:25885)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

This is also the same bug..
1. Settings -> User Interface -> Show Moon .. uncheck this option
2. Open the app at 11:59 PM and wait a minute

The app will crash at midnight with the following exception:
```
2020-05-03 00:00:00.705 32738-32738/com.forrestguice.suntimeswidget E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.forrestguice.suntimeswidget, PID: 32738
    java.lang.RuntimeException: Error receiving broadcast Intent { act=suntimes.SUNTIMES_APP_UPDATE_FULL flg=0x14 (has extras) } in com.forrestguice.suntimeswidget.SuntimesActivity$1@34fd3b3
        at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1401)
        at android.app.-$$Lambda$LoadedApk$ReceiverDispatcher$Args$_BumDX2UKsnxLVrE6UJsJZkotuA.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.forrestguice.suntimeswidget.calculator.SuntimesMoonData.invalidateCalculation()' on a null object reference
        at com.forrestguice.suntimeswidget.SuntimesActivity.invalidateData(SuntimesActivity.java:1449)
        at com.forrestguice.suntimeswidget.SuntimesActivity$1.onReceive(SuntimesActivity.java:507)
        at android.app.LoadedApk$ReceiverDispatcher$Args.lambda$getRunnable$0(LoadedApk.java:1391)
        at android.app.-$$Lambda$LoadedApk$ReceiverDispatcher$Args$_BumDX2UKsnxLVrE6UJsJZkotuA.run(Unknown Source:2) 
        at android.os.Handler.handleCallback(Handler.java:873) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6669) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
```

This one is similar but different (same root cause)..
1. Settings -> General -> Data Source -> sunrisesunsetlib
2. Settings -> User Interface -> Show Moon .. verify checked
3. Settings -> User Interface -> Show Solstice/Equinox .. uncheck
4. Set Location -> Current (last known) -> update

The app crashes with the following exception:
```
2020-05-03 10:25:54.350 2744-2744/com.forrestguice.suntimeswidget E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.forrestguice.suntimeswidget, PID: 2744
    java.lang.NullPointerException: Attempt to invoke virtual method 'void com.forrestguice.suntimeswidget.calculator.SuntimesEquinoxSolsticeDataset.invalidateCalculation()' on a null object reference
        at com.forrestguice.suntimeswidget.SuntimesActivity.invalidateData(SuntimesActivity.java:1450)
        at com.forrestguice.suntimeswidget.SuntimesActivity$8.onStart(SuntimesActivity.java:840)
        at com.forrestguice.suntimeswidget.getfix.GetFixTask.onPreExecute(GetFixTask.java:165)
        at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:648)
        at com.forrestguice.suntimeswidget.getfix.GetFixTask.executeTask(GetFixTask.java:64)
        at com.forrestguice.suntimeswidget.getfix.GetFixHelper.getFix(GetFixHelper.java:126)
        at com.forrestguice.suntimeswidget.SuntimesActivity.refreshLocation(SuntimesActivity.java:1208)
        at com.forrestguice.suntimeswidget.SuntimesActivity.onOptionsItemSelected(SuntimesActivity.java:1126)
        at android.app.Activity.onMenuItemSelected(Activity.java:3543)
        at android.support.v4.app.FragmentActivity.onMenuItemSelected(FragmentActivity.java:408)
        at android.support.v7.app.AppCompatActivity.onMenuItemSelected(AppCompatActivity.java:195)
        at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:113)
        at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected(WindowCallbackWrapper.java:113)
        at android.support.v7.app.ToolbarActionBar$2.onMenuItemClick(ToolbarActionBar.java:69)
        at android.support.v7.widget.Toolbar$1.onMenuItemClick(Toolbar.java:206)
        at android.support.v7.widget.ActionMenuView$MenuBuilderCallback.onMenuItemSelected(ActionMenuView.java:776)
        at android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:822)
        at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:156)
        at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:969)
        at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:959)
        at android.support.v7.widget.ActionMenuView.invokeItem(ActionMenuView.java:623)
        at android.support.v7.view.menu.ActionMenuItemView.onClick(ActionMenuItemView.java:154)
        at android.view.View.performClick(View.java:6597)
        at android.view.View.performClickInternal(View.java:6574)
        at android.view.View.access$3100(View.java:778)
        at android.view.View$PerformClick.run(View.java:25885)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

These exceptions were introduced in `v0.12.0` during refactor to RecyclerView.

--- a/app/src/androidTest/java/com/forrestguice/suntimeswidget/SuntimesActivityTest.java
+++ b/app/src/androidTest/java/com/forrestguice/suntimeswidget/SuntimesActivityTest.java
@@ -18,7 +18,12 @@
 
 package com.forrestguice.suntimeswidget;
 
+import android.app.Activity;
+import android.app.PendingIntent;
 import android.content.Context;
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
 import android.support.test.espresso.IdlingPolicies;
 import android.support.test.espresso.IdlingResource;
 import android.support.test.espresso.matcher.ViewMatchers;
@@ -582,5 +587,76 @@ public class SuntimesActivityTest extends SuntimesActivityTestBase
         timeFields.add( allOf(withId(R.id.text_time_sunset_nautical), isDescendantOfA(card), hasSibling(sibling)) );
         timeFields.add( allOf(withId(R.id.text_time_sunset_astro), isDescendantOfA(card), hasSibling(sibling)) );
         return timeFields;
+    }
+
+    @Test
+    public void test_partialUpdateReciever()
+    {
+        // test PendingIntent
+        final SuntimesActivity activity = (SuntimesActivity)activityRule.getActivity();
+        PendingIntent partialUpdateIntent = activity.getPartialUpdateIntent(activity);
+        try {
+            partialUpdateIntent.send();
+
+        } catch (PendingIntent.CanceledException e) {
+            e.printStackTrace();
+            fail("CanceledException!");
+        }
+
+        // test receiver
+        activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                activity.partialUpdateReceiver.onReceive(activity, new Intent(SuntimesActivity.SUNTIMES_APP_UPDATE_PARTIAL));
+                activity.finish();
+                assertTrue("app hasn't crashed", activity.isFinishing());
+            }
+        });
+    }
+
+    @Test
+    public void test_fullUpdateReciever()
+    {
+        // test PendingIntent
+        final SuntimesActivity activity = (SuntimesActivity)activityRule.getActivity();
+        PendingIntent fullUpdateIntent = activity.getFullUpdateIntent(activity);
+        try {
+            fullUpdateIntent.send();
+
+        } catch (PendingIntent.CanceledException e) {
+            e.printStackTrace();
+            fail("CanceledException!");
+        }
+
+        // test receiver
+        activity.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                activity.fullUpdateReceiver.onReceive(activity, new Intent(SuntimesActivity.SUNTIMES_APP_UPDATE_FULL));
+                activity.finish();
+                assertTrue("app hasn't crashed", activity.isFinishing());
+            }
+        });
+    }
+
+    @Test
+    public void test_issue408()
+    {
+        // test "show moon" enabled
+        SuntimesActivity activity = (SuntimesActivity)activityRule.getActivity();
+        SharedPreferences.Editor pref = PreferenceManager.getDefaultSharedPreferences(activity).edit();
+        pref.putBoolean(AppSettings.PREF_KEY_UI_SHOWMOON, true);
+        pref.commit();
+        activity.finish();
+        activityRule.launchActivity(activity.getIntent());
+        test_fullUpdateReciever();
+
+        // test "show moon" disabled
+        pref.putBoolean(AppSettings.PREF_KEY_UI_SHOWMOON, false);
+        pref.commit();
+        activity = (SuntimesActivity)activityRule.getActivity();
+        activity.finish();
+        activityRule.launchActivity(activity.getIntent());
+        test_fullUpdateReciever();
     }
 }

--- a/app/src/main/java/com/forrestguice/suntimeswidget/SuntimesActivity.java
+++ b/app/src/main/java/com/forrestguice/suntimeswidget/SuntimesActivity.java
@@ -495,7 +495,7 @@ public class SuntimesActivity extends AppCompatActivity
     {
         return PendingIntent.getBroadcast(context, 0, new Intent(SuntimesActivity.SUNTIMES_APP_UPDATE_FULL), 0);
     }
-    private BroadcastReceiver fullUpdateReceiver = new BroadcastReceiver()
+    protected BroadcastReceiver fullUpdateReceiver = new BroadcastReceiver()
     {
         @Override
         public void onReceive(Context context, Intent intent)
@@ -521,7 +521,7 @@ public class SuntimesActivity extends AppCompatActivity
     {
         return PendingIntent.getBroadcast(context, 0, new Intent(SuntimesActivity.SUNTIMES_APP_UPDATE_PARTIAL), 0);
     }
-    private BroadcastReceiver partialUpdateReceiver = new BroadcastReceiver()
+    protected BroadcastReceiver partialUpdateReceiver = new BroadcastReceiver()
     {
         @Override
         public void onReceive(Context context, Intent intent)

--- a/app/src/main/java/com/forrestguice/suntimeswidget/SuntimesActivity.java
+++ b/app/src/main/java/com/forrestguice/suntimeswidget/SuntimesActivity.java
@@ -1445,10 +1445,18 @@ public class SuntimesActivity extends AppCompatActivity
 
     protected void invalidateData( Context context )
     {
-        dataset.invalidateCalculation();
-        dataset_moon.invalidateCalculation();
-        dataset_equinox.invalidateCalculation();
-        card_adapter.invalidateData();
+        if (dataset != null) {
+            dataset.invalidateCalculation();
+        }
+        if (dataset_moon != null) {
+            dataset_moon.invalidateCalculation();
+        }
+        if (dataset_equinox != null) {
+            dataset_equinox.invalidateCalculation();
+        }
+        if (card_adapter != null) {
+            card_adapter.invalidateData();
+        }
         updateViews(context);
     }
 

